### PR TITLE
[qwen2-vl] fix FA2 inference

### LINKED ISSUE #39095: `Qwen2_5_VLVisionAttention` with flash attention has no `is_causal` attribute
### System Info

transformers v4.53.0

### Who can help?

@amyeroberts, @qubvel 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

New in v4.53.0, I see that Qwen2_5_VLVisionAttention can now use flash attention. 

https://github.com/huggingface/transformers/blob/v4.53.0/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L252

However, in this line https://github.com/huggingface/transformers/blob/v4.53.0/src/transformers/integrations/flash_attention.py#L71

we assume that Qwen2_5_VLVisionAttention has an is_causal attribute but by inspection it does not. This crashes inference!

### Expected behavior

I would expect it not to crash.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/39095 and set `mask=None` for FA2, otherwise inference fails because FA2 expects a 2d mask

Tested with `RUN_SLOW=1 pytest -k flash_attn` for all modified models

### CODE REVIEW

File: src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py [hunk 1]
```diff
@@ -659,7 +659,7 @@ def forward(
             cu_seqlens_k=cu_seqlens,
             max_seqlen_q=max_seqlen,
             max_seqlen_k=max_seqlen,
-            is_causal=False,
+            is_causal=self.is_causal,
```
Comment (vasqu) [inline src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py]: Is this really needed? Iirc, they always use the module's `is_causal` flag, the passed kwarg was for some weird BC behavior on sdpa

Comment (zucchini-nlp) [reply src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py]: Ah maybe you're right, now that I checked llama attention. Didn't know `self.is_causal` is for BC only. Though it is kinda weird, having to redfine `is_causal` in call which technically can be different from `self.is_causal` if not defined carefully

Comment (vasqu) [reply src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py]: At some point the kwarg option should be deprecated imo, we rely on the modules `is_causal` in most cases. But yea definitely weird atm :D

File: src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py [hunk 2]
```diff
@@ -634,14 +634,17 @@ def forward(
...
+        attention_mask = None
+        if self.config._attn_implementation != "flash_attention_2":
+            attention_mask = torch.full(
+                [1, 1, seq_length, key_states.shape[-2]],
+                torch.finfo(query_states.dtype).min,
+                device=query_states.device,
+                dtype=query_states.dtype,
+            )
+            for i in range(1, len(cu_seqlens)):
+                attention_mask[..., cu_seqlens[i - 1] : cu_seqlens[i], cu_seqlens[i - 1] : cu_seqlens[i]] = 0
```
Comment (vasqu) [inline src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py]: This probably also fixes #39067? Might be nice to change the flash attention integration path in the future to prioritize pos+cu_seq (even with an attention mask).

Comment (zucchini-nlp) [reply src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py]: No, it doesn't, that issue was there before refactor and I found it's related to FA2 using ragged inputs which the SDPA/eager path doesn't support in the same way

I will comment under #39067 when I have a clear fix. Let's keep this PR open until then, better to fix it once and forever

**Review by vasqu — APPROVED** [review]
LGTM, thanks (for bearing with me) :)

**Review by ArthurZucker — COMMENTED** [review]
I am not sure I understand why we create the mask in the attention layer, we should make sure we do it at the model level. it is the same for all layers

**Comment by zucchini-nlp** [comment]
I don't think there was a specific reason, it has been that way since the first model was released. In any case Qwen-VL models seem to be based on FA2 first, while other attn are added just because we wanted to support them

Thus the mask is prepared based on `cu_seqlens`

**Comment by JJJYmmm** [comment]
@zucchini-nlp  Hi, the FA2 may not be working as expected. This is because we use the following code in `_flash_attention_forward`:

https://github.com/huggingface/transformers/blob/ed36f8490eb3748f2424d854936cc6f816aeb486/src/transformers/modeling_flash_attention_utils.py#L423-L426

However, there is an issue with the variable names here, which causes  FA2 to not function correctly.

https://github.com/zucchini-nlp/transformers/blob/2cb7f21991478494c963ab8c4b52691d8d192b0d/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L266-L269

After correcting the code to the following version, I tested the model accuracy and it worked as expected 🤗

```python
        attn_output, _ = attention_interface(
            self,
            query_states,
            key_states,
            value_states,
            attention_mask=attention_mask,
            dropout=0.0,
            scaling=self.scaling,
            cu_seq_lens_k=cu_seqlens,  # pass cu seq lens for FA2
            cu_seq_lens_q=cu_seqlens,
            max_length_q=max_seqlen,
            max_length_k=max_seqlen,
            is_causal=False,
            **kwargs,
        )
```

**Comment by zucchini-nlp** [comment]
@JJJYmmm yeah, thanks, the current state doesn't work yet. In fact renaming only doesn't help because the `varlen` codepath needs position ids as well. I have the local version working and now I am trying to find the reason to a linked issue. Even though it's not caused by recent changes, we need to fix it

**Comment by JJJYmmm** [comment]
> @JJJYmmm yeah, thanks, the current state doesn't work yet. In fact renaming only doesn't help because the `varlen` codepath needs position ids as well. I have the local version working and now I am trying to find the reason to a linked issue. Even though it's not caused by recent changes, we need to fix it

Oh! Yes, I also modified the logic in `_flash_attention_forward` — when `position_ids` is `None` but `cu_seqlens` is not `None`, it now directly calls the varlen func. Looking forward to your changes!

**Comment by zucchini-nlp** [comment]
@vasqu @JJJYmmm I updated the fa2 path so it uses `flash_attn_with_varlen` for pad-free inputs. Can you check once again and if everything is fine, I'll merge

**Comment by JJJYmmm** [comment]
@zucchini-nlp  LGTM 🫡

File: src/transformers/modeling_flash_attention_utils.py [hunk 3]
```diff
@@ -508,6 +508,21 @@ def _flash_attention_forward(
...
+    # Case 2. Some models pass directly pre-computed `cu_seqlens` so we don't need to infer it from position ids. It is safe to
+    # use `flash_attn_varlen_func` knowing we already have all necessary the kwargs
+    is_fa2_with_position_ids = (
+        position_ids is not None
+        and query_states.shape[0] == 1
+        and (max_length_q is not None or (query_length != 1 and not (torch.diff(position_ids, dim=-1) >= 0).all()))
+    )
+    is_fa2_with_varlen_kwargs = all(
+        kwarg is not None for kwarg in (cu_seq_lens_q, cu_seq_lens_k, max_length_q, max_length_k)
+    )
```
Comment (vasqu) [inline src/transformers/modeling_flash_attention_utils.py:525]: I'm a bit unsure about this since we would allow `cu_seq` and `max_seq` only but on most models we also have RoPE so it's breaking those models silently if we eff up not passing `position_ids` (due to RoPE positions being bound to `position_ids` as well). We should imo add at least a warning on only varlen kwargs to give some discretion here. 

On another note, what do the integration tests use? Are they still working as expected :eyes: seems a bit sus

Comment (zucchini-nlp) [reply src/transformers/modeling_flash_attention_utils.py:525]: The integration tests always use `position_ids` which is the first `is_fa2_with_position_ids`, I just copied it from existing code and moved up here. The second case is added for Qwen only, afaik no other model passes pre-computed `cu_lens` form attention layers

In qwen we don't need any position ids, because they are 3D and won't help at all in inferring `cu_lens`

Comment (vasqu) [reply src/transformers/modeling_flash_attention_utils.py:525]: I meant the general integration tests like e.g. https://github.com/huggingface/transformers/blob/d53518c5f2dd7ada022ff5b725c684c9ed89cb44/tests/models/qwen2_vl/test_modeling_qwen2_vl.py#L507

For why, I'm concerned about the second case on qwen is future model additions and general usability, not the validity of qwen. For developers, `is_fa2_with_varlen_kwargs` indicates that this suffices for varlen - before, we (unintentionally) checked for the existence of (correct flattened) position ids that RoPE models need when using varlen. Maybe #35941 helps for reference on what I mean. 

Imo, it would help to add at least comments that for varlen most models need correct flattened position ids (from e.g. a collator), especially RoPE models which make up the majority of newer models.

Comment (zucchini-nlp) [reply src/transformers/modeling_flash_attention_utils.py:525]: Hmm, am I right that users might be passing only `cu_lens` without correct `position_ids`? I believe that would be users' responsibility to take care that RoPE is applied correctly, but I will add a comment in code explaining it, sure

In slow integration tests we don't pass `position_ids`, not that of I know. For most LLMs the fa2 path integration tests fallback to inferring `cu_lens` from the mask, and in Qwen the `position_ids` are constructed on-the-fly during forward call. The model has a requirement for adding rope deltas on top of 3D positions and I don't think users would be doing all that manually

Comment (vasqu) [reply src/transformers/modeling_flash_attention_utils.py:525]: "Hmm, am I right that users might be passing only cu_lens without correct position_ids?" - Yes, not only users but possibly us as well because it's something that's harder to figure out when done wrong imo :D

**Review by vasqu — APPROVED** [review]
Thanks again

**Comment by ArthurZucker** [comment]
@zucchini-nlp can we refactor the attention mask creation to be outside, cu seqs are not different for a layer!

File: src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py [hunk 4]
```diff
@@ -788,6 +779,25 @@ def get_input_embeddings(self) -> nn.Module:
...
+        seq_length = inputs_tensor.shape[1]
+        attention_mask = torch.full(
+            [1, 1, seq_length, seq_length],
+            torch.finfo(inputs_tensor.dtype).min,
+            device=inputs_tensor.device,
+            dtype=inputs_tensor.dtype,
+        )
+        for i in range(1, len(cu_seqlens)):
+            attention_mask[..., cu_seqlens[i - 1] : cu_seqlens[i], cu_seqlens[i - 1] : cu_seqlens[i]] = 0
+        return attention_mask
```
Comment (ArthurZucker) [inline src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py]:
```python
import torch
import time
import itertools
import random

def original_attention_mask(inputs_tensor, cu_seqlens):
    seq_length = inputs_tensor.shape[1]
    attention_mask = torch.full(
        [1, 1, seq_length, seq_length],
        torch.finfo(inputs_tensor.dtype).min,
        device=inputs_tensor.device,
        dtype=inputs_tensor.dtype,
    )
    for i in range(1, len(cu_seqlens)):
        attention_mask[..., cu_seqlens[i - 1]:cu_seqlens[i], cu_seqlens[i - 1]:cu_seqlens[i]] = 0
    return attention_mask

def vectorized_attention_mask(inputs_tensor, cu_seqlens):
    seq_length = inputs_tensor.shape[1]
    device = inputs_tensor.device
    dtype = inputs_tensor.dtype

    token_ids = torch.arange(seq_length, device=device)
    cu_seqlens = cu_seqlens.to(device)
    block_ids = torch.bucketize(token_ids, cu_seqlens[1:-1], right=False)

    block_mask = block_ids.unsqueeze(0) == block_ids.unsqueeze(1)
    attention_mask = torch.where(
        block_mask,
        torch.tensor(0.0, device=device, dtype=dtype),
        torch.finfo(dtype).min,
    )
    return attention_mask.unsqueeze(0).unsqueeze(0)

def generate_random_cu_seqlens(total_len, min_block=4, max_block=20):
    lengths = []
    remaining = total_len
    while remaining > 0:
        step = min(random.randint(min_block, max_block), remaining)
        lengths.append(step)
        remaining -= step
    cu_seqlens = [0]
    for l in lengths:
        cu_seqlens.append(cu_seqlens[-1] + l)
    return torch.tensor(cu_seqlens, dtype=torch.long)

def benchmark(fn, inputs_tensor, cu_seqlens, label, compile=False, warmup=5, runs=10):
    if compile:
        fn = torch.compile(fn)
        for _ in range(3):  # graph warmup
            fn(inputs_tensor, cu_seqlens)

    for _ in range(warmup):
        fn(inputs_tensor, cu_seqlens)

    torch.cuda.synchronize() if inputs_tensor.is_cuda else None
    start = time.time()
    for _ in range(runs):
        fn(inputs_tensor, cu_seqlens)
    torch.cuda.synchronize() if inputs_tensor.is_cuda else None
    duration = (time.time() - start) * 1000 / runs

    compile_str = " (compiled)" if compile else " (no compile)"
    print(f"{label:<30} |{compile_str:<15} B={inputs_tensor.shape[0]:<2} L={inputs_tensor.shape[1]:<4} D={inputs_tensor.shape[2]:<3} -> {duration:.3f} ms")

if __name__ == "__main__":
    torch.manual_seed(0)
    random.seed(0)
    device = "cuda" if torch.cuda.is_available() else "cpu"

    batch_sizes = [1, 4]
    seq_lengths = [128, 512, 1024]
    dims = [64, 128]

    for B, L, D in itertools.product(batch_sizes, seq_lengths, dims):
        inputs_tensor = torch.randn(B, L, D, device=device)
        cu_seqlens = generate_random_cu_seqlens(L).to(device)

        benchmark(original_attention_mask, inputs_tensor, cu_seqlens, "Original", compile=False)
        benchmark(original_attention_mask, inputs_tensor, cu_seqlens, "Original", compile=True)
        benchmark(vectorized_attention_mask, inputs_tensor, cu_seqlens, "Vectorized", compile=False)
        benchmark(vectorized_attention_mask, inputs_tensor, cu_seqlens, "Vectorized", compile=True)

```
I think this is ~2 as fast :)

**Review by ArthurZucker — APPROVED** [review]
Let's push!

File: src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py (hunk 4 above)
Comment (ArthurZucker) [reply src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py]: or not
image

**Comment by ArthurZucker** [comment]
image
Actually on cuda it makes a somewhat 2x diff
